File: php/tests/ClientTest.php
<?php
declare(strict_types=1);

namespace WPush\Tests;

use PHPUnit\Framework\TestCase;
use WPush\Client;
use WPush\HttpTransport;
use WPush\ValidationException;
use WPush\WPushException;

final class CapturingTransport implements HttpTransport
{
    public $lastUrl;
    public $lastHeaders;
    public $lastBody;
    public $response;

    public function __construct(array $response)
    {
        $this->response = $response;
    }

    public function post(string $url, array $headers, string $jsonBody): array
    {
        $this->lastUrl = $url;
        $this->lastHeaders = $headers;
        $this->lastBody = $jsonBody;
        return $this->response;
    }
}

final class ClientTest extends TestCase
{
    private function client(CapturingTransport $t): Client
    {
        return new Client('k', Client::DEFAULT_BASE_URL, 30.0, null, $t);
    }

    public function testSendSuccess(): void
    {
        $t = new CapturingTransport(['status' => 200, 'body' => json_encode(['code' => 0, 'message' => 'success', 'data' => '12345'])]);
        $mid = $this->client($t)->send('t', 'c', ['wechat', 'dingtalk']);
        $this->assertSame('12345', $mid);
        $this->assertSame(Client::USER_AGENT, $t->lastHeaders['User-Agent']);
        $this->assertSame('k', $t->lastHeaders['X-API-Key']);
        $body = json_decode($t->lastBody, true);
        $this->assertSame('k', $body['apikey']);
        $this->assertSame('wechat,dingtalk', $body['channel']);
        $this->assertStringEndsWith('/api/v1/send', $t->lastUrl);
    }

    public function testSendNumericDataStringified(): void
    {
        $t = new CapturingTransport(['status' => 200, 'body' => json_encode(['code' => 0, 'message' => 'ok', 'data' => 999])]);
        $this->assertSame('999', $this->client($t)->send('t'));
    }

    public function testRejectBoolCode(): void
    {
        $t = new CapturingTransport(['status' => 200, 'body' => json_encode(['code' => false, 'message' => 'nope', 'data' => null])]);
        try {
            $this->client($t)->send('t');
            $this->fail('expected exception');
        } catch (WPushException $e) {
            $this->assertNull($e->getErrorCode());
        }
    }

    public function testApiError401(): void
    {
        $t = new CapturingTransport(['status' => 200, 'body' => json_encode(['code' => 401, 'message' => 'API Key错误', 'data' => null])]);
        try {
            $this->client($t)->send('t');
            $this->fail('expected');
        } catch (WPushException $e) {
            $this->assertSame(401, $e->getErrorCode());
        }
    }

    public function testOptionTopicConflict(): void
    {
        $t = new CapturingTransport(['status' => 200, 'body' => '{}']);
        $this->expectException(ValidationException::class);
        $this->client($t)->send('t', null, null, 'ops', 'topic1');
        $this->assertNull($t->lastUrl);
    }

    public function testMissingApiKey(): void
    {
        $prev = getenv('WPUSH_API_KEY');
        putenv('WPUSH_API_KEY');
        try {
            $this->expectException(ValidationException::class);
            new Client();
        } finally {
            if ($prev === false) {
                putenv('WPUSH_API_KEY');
            } else {
                putenv('WPUSH_API_KEY=' . $prev);
            }
        }
    }

    public function testQuerySuccess(): void
    {
        $data = ['id' => '123', 'status' => 1, 'title' => 't'];
        $t = new CapturingTransport(['status' => 200, 'body' => json_encode(['code' => 0, 'message' => 'success', 'data' => $data])]);
        $out = $this->client($t)->query('123');
        $this->assertSame('123', $out['id']);
        $this->assertSame(1, $out['status']);
        $body = json_decode($t->lastBody, true);
        $this->assertSame('123', $body['id']);
    }

    public function testEnvApiKey(): void
    {
        putenv('WPUSH_API_KEY=from-env');
        try {
            $c = new Client(null, Client::DEFAULT_BASE_URL, 30.0, null, new CapturingTransport(['status' => 200, 'body' => '{}']));
            $this->assertSame('from-env', $c->getApiKey());
        } finally {
            putenv('WPUSH_API_KEY');
        }
    }

    public function testIdempotencyHeader(): void
    {
        $t = new CapturingTransport(['status' => 200, 'body' => json_encode(['code' => 0, 'message' => 'ok', 'data' => '1'])]);
        $this->client($t)->send('t', null, null, null, null, null, 'idem-1');
        $this->assertSame('idem-1', $t->lastHeaders['X-Idempotency-Key']);
    }

    public function testSendMailSuccess(): void
    {
        $t = new CapturingTransport(['status' => 200, 'body' => json_encode(['code' => 0, 'message' => 'success', 'data' => 'mail-1'])]);
        $this->assertSame('mail-1', $this->client($t)->sendMail('a@b.com', 'hi', 'body'));
        $this->assertStringEndsWith('/api/v1/send_mail', $t->lastUrl);
        $body = json_decode($t->lastBody, true);
        $this->assertSame('a@b.com', $body['to']);
    }

    public function testSendCodeSuccess(): void
    {
        $t = new CapturingTransport(['status' => 200, 'body' => json_encode(['code' => 0, 'message' => 'success', 'data' => 'code-1'])]);
        $this->assertSame('code-1', $this->client($t)->sendCode('13800138000', '123456'));
        $this->assertStringEndsWith('/api/v1/send_code', $t->lastUrl);
    }

    public function testSendCodeBadPhone(): void
    {
        $t = new CapturingTransport(['status' => 200, 'body' => '{}']);
        $this->expectException(ValidationException::class);
        $this->client($t)->sendCode('123', '123456');
    }

    public function testQueryRelaySuccess(): void
    {
        $t = new CapturingTransport(['status' => 200, 'body' => json_encode(['code' => 0, 'message' => 'success', 'data' => ['id' => 'r1', 'status' => 1]])]);
        $out = $this->client($t)->queryRelay('r1');
        $this->assertSame('r1', $out['id']);
        $this->assertStringEndsWith('/api/v1/query_relay', $t->lastUrl);
    }
}

File: java/src/test/java/cn/wpush/sdk/ClientTest.java
package cn.wpush.sdk;

import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class ClientTest {

  static final class CapturingTransport implements HttpTransport {
    String lastUrl;
    Map<String, String> lastHeaders;
    String lastBody;
    Response next;

    CapturingTransport(Response next) {
      this.next = next;
    }

    @Override
    public Response post(String url, Map<String, String> headers, String jsonBody) {
      this.lastUrl = url;
      this.lastHeaders = headers;
      this.lastBody = jsonBody;
      return next;
    }
  }

  private static Client client(HttpTransport t) {
    return Client.builder().apiKey("k").transport(t).build();
  }

  @Test
  void sendSuccessJoinsChannelAndHeaders() {
    CapturingTransport t = new CapturingTransport(
        new HttpTransport.Response(200, "{\"code\":0,\"message\":\"success\",\"data\":\"12345\"}"));
    Client c = client(t);
    String mid = c.send("t", "c", Arrays.asList("wechat", "dingtalk"), null, null, null, null);
    assertEquals("12345", mid);
    assertEquals(Client.USER_AGENT, t.lastHeaders.get("User-Agent"));
    assertEquals("k", t.lastHeaders.get("X-API-Key"));
    assertTrue(t.lastUrl.endsWith("/api/v1/send"));
    Map<String, Object> body = Json.parseObject(t.lastBody);
    assertEquals("k", body.get("apikey"));
    assertEquals("wechat,dingtalk", body.get("channel"));
    assertEquals("t", body.get("title"));
  }

  @Test
  void sendNumericDataStringified() {
    CapturingTransport t = new CapturingTransport(
        new HttpTransport.Response(200, "{\"code\":0,\"message\":\"ok\",\"data\":999}"));
    assertEquals("999", client(t).send("t"));
  }

  @Test
  void rejectBoolCode() {
    CapturingTransport t = new CapturingTransport(
        new HttpTransport.Response(200, "{\"code\":false,\"message\":\"nope\",\"data\":null}"));
    WPushException ex = assertThrows(WPushException.class, () -> client(t).send("t"));
    assertNull(ex.getCode());
  }

  @Test
  void apiError401() {
    CapturingTransport t = new CapturingTransport(
        new HttpTransport.Response(200, "{\"code\":401,\"message\":\"API Key错误\",\"data\":null}"));
    WPushException ex = assertThrows(WPushException.class, () -> client(t).send("t"));
    assertEquals(401, ex.getCode());
  }

  @Test
  void optionTopicConflict() {
    CapturingTransport t = new CapturingTransport(new HttpTransport.Response(200, "{}"));
    Client c2 = client(t);
    assertThrows(ValidationException.class, () -> c2.send("t", null, null, "ops", "topic1", null, null));
    assertNull(t.lastUrl);
  }

  @Test
  void missingApiKey() {
    String prev = System.getenv("WPUSH_API_KEY");
    // Cannot easily clear env; pass empty and ensure no env dependency by using blank after checking
    // If env is set in CI this may pass; use builder with explicit empty only when env unset.
    if (prev == null || prev.isEmpty()) {
      assertThrows(ValidationException.class, () -> Client.builder().build());
    }
  }

  @Test
  void querySuccess() {
    CapturingTransport t = new CapturingTransport(
        new HttpTransport.Response(200, "{\"code\":0,\"message\":\"success\",\"data\":{\"id\":\"123\",\"status\":1,\"title\":\"t\"}}"));
    Map<String, Object> out = client(t).query("123");
    assertEquals("123", out.get("id"));
    assertEquals(1, out.get("status"));
    Map<String, Object> body = Json.parseObject(t.lastBody);
    assertEquals("123", body.get("id"));
    assertEquals("k", body.get("apikey"));
  }

  @Test
  void idempotencyHeader() {
    CapturingTransport t = new CapturingTransport(
        new HttpTransport.Response(200, "{\"code\":0,\"message\":\"ok\",\"data\":\"1\"}"));
    client(t).send("t", null, null, null, null, null, "idem-1");
    assertEquals("idem-1", t.lastHeaders.get("X-Idempotency-Key"));
  }

  @Test
  void sendMailSuccess() {
    CapturingTransport t = new CapturingTransport(
        new HttpTransport.Response(200, "{\"code\":0,\"message\":\"success\",\"data\":\"mail-1\"}"));
    assertEquals("mail-1", client(t).sendMail("a@b.com", "hi", "body"));
    assertTrue(t.lastUrl.endsWith("/api/v1/send_mail"));
    Map<String, Object> body = Json.parseObject(t.lastBody);
    assertEquals("a@b.com", body.get("to"));
    assertEquals("hi", body.get("title"));
    assertEquals("body", body.get("content"));
  }

  @Test
  void sendCodeSuccess() {
    CapturingTransport t = new CapturingTransport(
        new HttpTransport.Response(200, "{\"code\":0,\"message\":\"success\",\"data\":\"code-1\"}"));
    assertEquals("code-1", client(t).sendCode("13800138000", "123456"));
    assertTrue(t.lastUrl.endsWith("/api/v1/send_code"));
    Map<String, Object> body = Json.parseObject(t.lastBody);
    assertEquals("13800138000", body.get("phone"));
    assertEquals("123456", body.get("code"));
  }

  @Test
  void sendCodeBadPhone() {
    CapturingTransport t = new CapturingTransport(new HttpTransport.Response(200, "{}"));
    assertThrows(ValidationException.class, () -> client(t).sendCode("123", "123456"));
    assertNull(t.lastUrl);
  }

  @Test
  void queryRelaySuccess() {
    CapturingTransport t = new CapturingTransport(
        new HttpTransport.Response(200, "{\"code\":0,\"message\":\"success\",\"data\":{\"id\":\"r1\",\"status\":1}}"));
    Map<String, Object> out = client(t).queryRelay("r1");
    assertEquals("r1", out.get("id"));
    assertTrue(t.lastUrl.endsWith("/api/v1/query_relay"));
    Map<String, Object> body = Json.parseObject(t.lastBody);
    assertEquals("r1", body.get("id"));
  }

  @Test
  void isSuccessCodeRejectsBool() {
    assertFalse(Client.isSuccessCode(Boolean.FALSE));
    assertFalse(Client.isSuccessCode(Boolean.TRUE));
    assertTrue(Client.isSuccessCode(0));
    assertTrue(Client.isSuccessCode(0.0));
  }
}
